xds/internal/resolver/xds_resolver_test.go

```diff
@@ -1178,295 +1265,349 @@ func (s) TestXDSResolverMaxStreamDuration(t *testing.T) {
+	val, err := tcc.stateCh.Receive(ctx)
 	if err != nil {
-		t.Fatalf("Error waiting for UpdateState to be called: %v", err)
+		t.Fatalf("Timeout waiting for an update from the resolver: %v", err)
 	}
-	rState := gotState.(resolver.State)
+	rState := val.(resolver.State)
 	if err := rState.ServiceConfig.Err; err != nil {
-		t.Fatalf("ClientConn.UpdateState received error in service config: %v", rState.ServiceConfig.Err)
+		t.Fatalf("Received error in service config: %v", rState.ServiceConfig.Err)
 	}
-
-	wantJSON := `{"loadBalancingConfig":[{
-	 "xds_cluster_manager_experimental":{
-	   "children":{
-		 "cluster:test-cluster-1":{
-		   "childPolicy":[{"cds_experimental":{"cluster":"test-cluster-1"}}]
-		 }
-	   }
-	 }}]}`
-	wantSCParsed := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(wantJSON)
+	wantSCParsed := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(`
+{
+	"loadBalancingConfig": [
+		{
+		  "xds_cluster_manager_experimental": {
+			"children": {
+			  "cluster:OLD-CLUSTER": {
+				"childPolicy": [
+				  {
+					"cds_experimental": {
+					  "cluster": "OLD-CLUSTER"
+					} 
+				  } 
+				] 
+			  } 
+			} 
+		  } 
+		} 
+	  ] 
+}`)
 	if !internal.EqualServiceConfigForTesting(rState.ServiceConfig.Config, wantSCParsed.Config) {
-		t.Errorf("ClientConn.UpdateState received different service config")
+		t.Errorf("Received unexpected service config")
 		t.Error("got: ", cmp.Diff(nil, rState.ServiceConfig.Config))
 		t.Fatal("want: ", cmp.Diff(nil, wantSCParsed.Config))
 	}
 
+	// Make an RPC, but do not commit it yet.
 	cs := iresolver.GetConfigSelector(rState)
 	if cs == nil {
-		t.Fatal("received nil config selector")
+		t.Fatal("Received nil config selector in update from resolver")
 	}
-
-	res, err := cs.SelectConfig(iresolver.RPCInfo{Context: context.Background()})
+	resOld, err := cs.SelectConfig(iresolver.RPCInfo{Context: ctx, Method: "/service/method"})
 	if err != nil {
-		t.Fatalf("Unexpected error from cs.SelectConfig(_): %v", err)
+		t.Fatalf("cs.SelectConfig(): %v", err)
 	}
-	cluster := clustermanager.GetPickedClusterForTesting(res.Context)
-	if cluster != "cluster:test-cluster-1" {
-		t.Fatalf("")
+	if cluster := clustermanager.GetPickedClusterForTesting(resOld.Context); cluster != "cluster:OLD-CLUSTER" {
+		t.Fatalf("Picked cluster is %q, want %q", cluster, "cluster:OLD-CLUSTER")
 	}
-	// delay res.OnCommitted()
-
-	// Perform TWO updates to ensure the old config selector does not hold a
-	// reference to test-cluster-1.
-	xdsC.InvokeWatchRouteConfigCallback("", xdsresource.RouteConfigUpdate{
-		VirtualHosts: []*xdsresource.VirtualHost{
-			{
-				Domains: []string{targetStr},
-				Routes:  []*xdsresource.Route{{Prefix: newStringP(""), WeightedClusters: map[string]xdsresource.WeightedCluster{"NEW": {Weight: 1}}}},
-			},
-		},
-	}, nil)
-	tcc.stateCh.Receive(ctx) // Ignore the first update.
+	// delay resOld.OnCommitted()
```

Comment:
Could you add a note as to why this delay would keep old cluster around even if not specified in new resource? I.e. pending RPCs need to continue to proceed as normal etc.